Sprint 1

ArtMikhailov opened the pull request.

ArtMikhailov committed:
CustomContainer and CustomContainerCompose are implemented

ArturNurtdinov commented on app/src/main/java/com/example/androidpracticumcustomview/ui/theme/CustomContainer.kt:
```diff
@@ -1,40 +1,139 @@
+        bottom: Int
+    ) {
+        super.onLayout(changed, left, top, right, bottom)
+        if (childCount > 2) {
+            throw IllegalStateException("CustomContainer can only have two child views.")
+        }
```
Данную проверку лучше в `addView` сделать, чтобы при попытке добавить проверялось условие, а не при размешении вью, т.к. эти два процесса могут идти с большим временным промежутком

ArtMikhailov replied:
Я сделал ее в onLayout чтобы ограничение на кол-во вью работало и в случае если CustomContainer добавляется в xml верстке и в ней же прописываются дочерние вью. Но и в addView данная проверка тоже есть, чтобы в случае программного добавления вью исключение выбрасывалось сразу в месте вызова.

ArturNurtdinov commented on app/src/main/java/com/example/androidpracticumcustomview/ui/theme/CustomContainer.kt:
```diff
@@ -1,40 +1,139 @@
+            // we need to consider child at index 1 as well for placing at top
+            val viewToPlaceAtTop = if (childZero != null && childZero != bottomView) {
+                childZero
+            } else childOne
```
Такой корнер кейс лучше не обрабатывать, т.к. такая ситуация извне должна быть пофикшена, а не в этой вью

ArtMikhailov replied:
Убрал ) действительно какой то оверинжиниринг получился с этим кейсом.

ArturNurtdinov commented on app/src/main/java/com/example/androidpracticumcustomview/ui/theme/CustomContainer.kt:
```diff
@@ -1,40 +1,139 @@
+    private fun animateView(view: View, translationYBy: Float) {
+        view.animate()
+            .alpha(1.0f)
+            .setDuration(2000)
```
Значения задержки лучше вынести в константы для повышения читабельности кода

ArtMikhailov replied:
Сделал

ArtMikhailov committed:
CustomContainer changes:
- durations of animations moved to constants
- removed cumbersome logic of handling corner case from CustomContainer
